signature/signature.go

```diff
@@ -199,6 +200,95 @@ func (s untrustedSignature) sign(mech SigningMechanism, keyIdentity string) ([]b
 	return mech.Sign(json, keyIdentity)
 }
 
+// cosignSignature is the payload format for sigstore signing an image using
+// the simple signing format.
+type cosignSignature struct {
+	DockerManifestDigest digest.Digest
+	DockerReference      string
+}
+
+// newCosignSignature returns a cosignSignature object with
+// the specified primary contents and appropriate metadata.
+func NewCosignSignature(dockerManifestDigest digest.Digest, dockerReference string) cosignSignature {
+	// Use intermediate variables for these values so that we can take their addresses.
+	// Golang guarantees that they will have a new address on every execution.
+	return cosignSignature{
+		DockerManifestDigest: dockerManifestDigest,
+		DockerReference:      dockerReference,
+	}
+}
+
+// Compile-time check that cosignSignature implements json.Marshaler
+var _ json.Marshaler = (*cosignSignature)(nil)
+
+// MarshalJSON implements the json.Marshaler interface.
+func (s cosignSignature) MarshalJSON() ([]byte, error) {
+	if s.DockerManifestDigest == "" || s.DockerReference == "" {
+		return nil, errors.New("Unexpected empty signature content")
+	}
+	critical := map[string]interface{}{
+		"type":     cosignSignatureType,
+		"image":    map[string]string{"docker-manifest-digest": s.DockerManifestDigest.String()},
+		"identity": map[string]string{"docker-reference": s.DockerReference},
+	}
+	optional := map[string]interface{}{} // TODO: add annotations i.e. claims
```

Comment:
I do think there’s non-trivial long-term value with including the name and version number of the producer inside the signature, so that future versions can add workarounds for known bugs in old software.